src/run/run.c

```diff
@@ -1720,17 +1720,29 @@ static int start_transient_service(sd_bus *bus) {
                         if (r < 0)
                                 return log_error_errno(r, "Failed to get machine PTY: %s", bus_error_message(&error, r));
 
-                        r = sd_bus_message_read(pty_reply, "hs", &master, &s);
+                        r = sd_bus_message_read(pty_reply, "hs", &master, &pty_path);
                         if (r < 0)
                                 return bus_log_parse_error(r);
 
                         master = fcntl(master, F_DUPFD_CLOEXEC, 3);
                         if (master < 0)
                                 return log_error_errno(errno, "Failed to duplicate master fd: %m");
 
-                        pty_path = strdup(s);
-                        if (!pty_path)
-                                return log_oom();
+                        r = sd_bus_get_property_trivial(
+                                        system_bus,
+                                        "org.freedesktop.machine1",
+                                        strjoina("/org/freedesktop/machine1/machine/", arg_host),
+                                        "org.freedesktop.machine1.Machine",
+                                        "Leader",
+                                        &error,
+                                        'u', &leader);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to get machine leader: %s", bus_error_message(&error, r));
+
+                        pty_slave = open_terminal_in_namespace(leader, pty_path, O_RDWR|O_NOCTTY|O_CLOEXEC);
```

Comment:
Hmm, I think we shouldn't bypass machined here. It would result in a weird mix if the pty is created by machined yet systemd-run directly enter the ns to acquire the fd. Instead, we would likely need a `OpenMachinePTYEx` method that returns the pty fd and potentially `chmod()` the pty (also see https://github.com/systemd/systemd/pull/32999#discussion_r1612925400).

Hence for v256, please pass the plain `TTYPath=` only if --machine is used. I'll revisit this in v257.

Reply:
I started adding a new method that returned the slave pty fd but finally changed my mind: adding a new method at this point (close to the release of v256) wouldn't be a good idea. Also as said, a fix needs to be backported to the stable trees, so better to keep the fix as simple as possible.

Reply:
I'll update the patch to pass `TTYPath=` again.

Reply:
yes, please do not bypass machined, let's do that in machined.

Reply:
You meant adding the new method before v256 release ?

Reply:
yeah

Reply:
@YHNdnzj are you already on it ?